### PR TITLE
Display units: line lists

### LINKED ISSUE #2028: Update line lists to use display unit
**Reporter: Jesse Averbukh**

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request updates the line lists plugin so that existing lines are unit-aware.  This does not yet update the display of lines in the plugin itself to show in the display units - that work will likely be deferred for the refactor of the line list plugin as both of those efforts will require changing the way the information is stored internally.

https://user-images.githubusercontent.com/877591/231570983-c7be856b-bba2-4d16-9228-41482e4e735d.mov

Changelog will point to https://github.com/spacetelescope/jdaviz/pull/2127 so is ignored from CI here.

Closes #2028 

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Review by javerbukh — APPROVED** [review]
I tried adding custom lines after changing the units and also adding custom lines in different units and then switching to those units. Everything worked as expected. I did notice that the table at the bottom of the plugin showing rest and observed units could potentially be updated if it made scientific sense. Should the observed value be in the original units of the line or should it update with the spectral unit? Since I don't know if that is a valid question and everything else looks good I'll go ahead and approve!

**Comment by kecnry** [comment]
> I did notice that the table at the bottom of the plugin showing rest and observed units could potentially be updated if it made scientific sense. Should the observed value be in the original units of the line or should it update with the spectral unit?

I think we'll eventually want this (at least as the default), but I think it makes sense to defer until the refactor (which will make use of a database of all lines, etc).

**Comment by kecnry** [comment]
putting this back in draft until CI can be fixed now that this is merged on top of #2132

**Comment by rosteen** [comment]
Thought I had approved this, sorry. I tested and saw that everything worked on Friday, let me do one more pass of the actual code before I give it a check mark.

File: jdaviz/core/marks.py [hunk 1]
```diff
@@ -59,11 +59,79 @@ def _update_counts(self, *args):
         self.right.text = [f'{oob_right} \u25b6' if oob_right > 0 else '']
 
 
-class BaseSpectrumVerticalLine(Lines, HubListener):
+class PluginMark():
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
Comment (rosteen) [inline jdaviz/core/marks.py:64]: Wait, this doesn't inherit anything...does the base python `Class` init handle setting `self.[whatever]` for kwargs if you call `super().__init__` like this?

File: jdaviz/core/marks.py [hunk 2]
```diff
@@ -59,11 +59,79 @@ def _update_counts(self, *args):
...
+
+    def update_xy(self, x, y):
+        self.x = np.asarray(x)
+        self.y = np.asarray(y)
+
+    def append_xy(self, x, y):
+        self.x = np.append(self.x, x)
+        self.y = np.append(self.y, y)
+
+    def _update_units(self):
```
Comment (rosteen) [inline jdaviz/core/marks.py:85]: Does this ever get called other than during `__init__`?

**Comment by rosteen** [comment]
I removed the other reviewer tags since I'm 2, I think that's at least in the direction of what we agreed on for avoiding overloading reviews 😃

File: jdaviz/core/marks.py (hunk 1 above)
Comment (kecnry) [reply jdaviz/core/marks.py:64]: turns out its required in order for the `comm` to be set correctly when other classes inherit from this and `Lines`

File: jdaviz/core/marks.py [hunk 3]
```diff
@@ -59,11 +59,77 @@ def _update_counts(self, *args):
...
+        self.auto_update_units = True
+        self.hub.subscribe(self, GlobalDisplayUnitChanged,
+                           handler=self._on_global_display_unit_changed)
+
+        if not self.auto_update_units:
+            return
+        if self.xunit is None:
+            self.set_x_unit()
+        if self.yunit is None:
+            self.set_y_unit()
```
Comment (rosteen) [inline jdaviz/core/marks.py:77]:
```suggestion
        if self.auto_update_units:
            if self.xunit is None:
                self.set_x_unit()
            if self.yunit is None:
                self.set_y_unit()
```

I'm not sure if the if statement is even needed now here (I doubt the plugin changes the `auto_update_units` value before `__init__` finishes), but if it is needed this is slightly shorter.

**Pull request merged**